### ISSUE TITLE
(machine_charm) Labels are not applied to metrics gathered through the cos_machine relation

### ISSUE THREAD

**Issue by dstathis** [canonical/grafana-agent-k8s-operator#108]
### Bug Description

When scraping over the cos_machine relation, labels are not assigned to the metrics.

### To Reproduce

1. Deploy kafka or zookeeper
2. Deploy grafana-agent and relate to the above charm
3. Relate to prometheus via cross-model relation
4. Check the labels in the prometheus UI (for example the "up" query)

### Environment

-

### Relevant log output

```shell
-
```

### Additional context

![Screenshot from 2023-03-07 16-09-00](https://user-images.githubusercontent.com/2110777/223505156-4e1acde4-e909-4f0d-99e6-d487c89ecae2.png)

**Comment by sed-i** [comment]
Using charm from commit [556d999](https://github.com/canonical/grafana-agent-k8s-operator/commit/556d9993d7d9e5061de7158552396889556657a8).

Labels do appear in gagent's target for kafka:
```terminal
ubuntu@juju-0728c2-4:~$ curl -s localhost:12345/agent/api/v1/metrics/targets | jq
```
```json
{
  "status": "success",
  "data": [
    {
      "instance": "8a263e2cb5aa14d547a83233b446c3bb",
      "target_group": "integrations/agent",
      "endpoint": "http://127.0.0.1:12345/integrations/agent/metrics",
      "state": "up",
      "labels": {
        "agent_hostname": "juju-0728c2-4",
        "instance": "default_863a1762-9849-4724-8f72-e2288d0728c2_kafka_kafka/2",
        "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_self-monitoring",
        "juju_application": "agent",
        "juju_charm": "grafana-agent",
        "juju_model": "default",
        "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
        "juju_unit": "agent/6"
      },
      "discovered_labels": {
        "__address__": "127.0.0.1:12345",
        "__metrics_path__": "/integrations/agent/metrics",
        "__scheme__": "http",
        "__scrape_interval__": "1m",
        "__scrape_timeout__": "10s",
        "agent_hostname": "juju-0728c2-4",
        "job": "integrations/agent"
      },
      "last_scrape": "2023-03-07T20:01:47.324327671Z",
      "scrape_duration_ms": 4,
      "scrape_error": ""
    },
    {
      "instance": "8a263e2cb5aa14d547a83233b446c3bb",
      "target_group": "integrations/node_exporter",
      "endpoint": "http://127.0.0.1:12345/integrations/node_exporter/metrics",
      "state": "up",
      "labels": {
        "agent_hostname": "juju-0728c2-4",
        "instance": "default_863a1762-9849-4724-8f72-e2288d0728c2_kafka_kafka/2",
        "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_node-exporter",
        "juju_application": "kafka",
        "juju_model": "default",
        "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
        "juju_unit": "kafka/2"
      },
      "discovered_labels": {
        "__address__": "127.0.0.1:12345",
        "__metrics_path__": "/integrations/node_exporter/metrics",
        "__scheme__": "http",
        "__scrape_interval__": "1m",
        "__scrape_timeout__": "10s",
        "agent_hostname": "juju-0728c2-4",
        "job": "integrations/node_exporter"
      },
      "last_scrape": "2023-03-07T20:01:11.780799504Z",
      "scrape_duration_ms": 8,
      "scrape_error": ""
    }
  ]
}
```

And the target does have metrics:
```terminal
ubuntu@juju-0728c2-4:~$ curl -s http://127.0.0.1:12345/integrations/node_exporter/metrics |head
# HELP node_arp_entries ARP entries by device
# TYPE node_arp_entries gauge
node_arp_entries{device="eth0"} 3
# HELP node_boot_time_seconds Node boot time, in unixtime.
# TYPE node_boot_time_seconds gauge
node_boot_time_seconds 1.678124432e+09
```

And querying prometheus does show the labels:
```terminal
$ curl -s --data-urlencode 'match[]={__name__=~"node_arp_entries"}' http://10.43.8.34:80/welcome-prom-0/api/v1/series | jq
```
```json
{
  "status": "success",
  "data": [
    {
      "__name__": "node_arp_entries",
      "agent_hostname": "juju-0728c2-4",
      "device": "eth0",
      "instance": "default_863a1762-9849-4724-8f72-e2288d0728c2_kafka_kafka/2",
      "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_node-exporter",
      "juju_application": "kafka",
      "juju_model": "default",
      "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
      "juju_unit": "kafka/2"
    }
  ]
}
```
Same for `up`.

Perhaps the issue got resolved in the commits in between? @dstathis

**Comment by dstathis** [comment]
> Using charm from commit [556d999](https://github.com/canonical/grafana-agent-k8s-operator/commit/556d9993d7d9e5061de7158552396889556657a8) over `juju-info`.
> 
> Labels do appear in gagent's target for kafka:
> ```terminal
> ubuntu@juju-0728c2-4:~$ curl -s localhost:12345/agent/api/v1/metrics/targets | jq
> ```
> ```json
> {
>   "status": "success",
>   "data": [
>     {
>       "instance": "8a263e2cb5aa14d547a83233b446c3bb",
>       "target_group": "integrations/agent",
>       "endpoint": "http://127.0.0.1:12345/integrations/agent/metrics",
>       "state": "up",
>       "labels": {
>         "agent_hostname": "juju-0728c2-4",
>         "instance": "default_863a1762-9849-4724-8f72-e2288d0728c2_kafka_kafka/2",
>         "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_self-monitoring",
>         "juju_application": "agent",
>         "juju_charm": "grafana-agent",
>         "juju_model": "default",
>         "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
>         "juju_unit": "agent/6"
>       },
>       "discovered_labels": {
>         "__address__": "127.0.0.1:12345",
>         "__metrics_path__": "/integrations/agent/metrics",
>         "__scheme__": "http",
>         "__scrape_interval__": "1m",
>         "__scrape_timeout__": "10s",
>         "agent_hostname": "juju-0728c2-4",
>         "job": "integrations/agent"
>       },
>       "last_scrape": "2023-03-07T20:01:47.324327671Z",
>       "scrape_duration_ms": 4,
>       "scrape_error": ""
>     },
>     {
>       "instance": "8a263e2cb5aa14d547a83233b446c3bb",
>       "target_group": "integrations/node_exporter",
>       "endpoint": "http://127.0.0.1:12345/integrations/node_exporter/metrics",
>       "state": "up",
>       "labels": {
>         "agent_hostname": "juju-0728c2-4",
>         "instance": "default_863a1762-9849-4724-8f72-e2288d0728c2_kafka_kafka/2",
>         "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_node-exporter",
>         "juju_application": "kafka",
>         "juju_model": "default",
>         "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
>         "juju_unit": "kafka/2"
>       },
>       "discovered_labels": {
>         "__address__": "127.0.0.1:12345",
>         "__metrics_path__": "/integrations/node_exporter/metrics",
>         "__scheme__": "http",
>         "__scrape_interval__": "1m",
>         "__scrape_timeout__": "10s",
>         "agent_hostname": "juju-0728c2-4",
>         "job": "integrations/node_exporter"
>       },
>       "last_scrape": "2023-03-07T20:01:11.780799504Z",
>       "scrape_duration_ms": 8,
>       "scrape_error": ""
>     }
>   ]
> }
> ```
> 
> And the target does have metrics:
> ```terminal
> ubuntu@juju-0728c2-4:~$ curl -s http://127.0.0.1:12345/integrations/node_exporter/metrics |head
> # HELP node_arp_entries ARP entries by device
> # TYPE node_arp_entries gauge
> node_arp_entries{device="eth0"} 3
> # HELP node_boot_time_seconds Node boot time, in unixtime.
> # TYPE node_boot_time_seconds gauge
> node_boot_time_seconds 1.678124432e+09
> ```
> 
> And querying prometheus does show the labels:
> ```terminal
> $ curl -s --data-urlencode 'match[]={__name__=~"node_arp_entries"}' http://10.43.8.34:80/welcome-prom-0/api/v1/series | jq
> ```
> ```json
> {
>   "status": "success",
>   "data": [
>     {
>       "__name__": "node_arp_entries",
>       "agent_hostname": "juju-0728c2-4",
>       "device": "eth0",
>       "instance": "default_863a1762-9849-4724-8f72-e2288d0728c2_kafka_kafka/2",
>       "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_node-exporter",
>       "juju_application": "kafka",
>       "juju_model": "default",
>       "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
>       "juju_unit": "kafka/2"
>     }
>   ]
> }
> ```
> Same for `up`.
> 
> Confirmed it's up to date with
> ```bash
> curl -X POST -g 'http://10.43.8.34:80/welcome-prom-0/api/v1/admin/tsdb/delete_series?match[]={__name__=~"node_arp_entries"}'
> 
> curl -s 'http://10.43.8.34:80/welcome-prom-0/api/v1/query?query=node_arp_entries' | jq
> ```
> 
> Perhaps the issue got resolved in the commits in between? @dstathis 

I don't see Kafka being scraped anywhere in that config. I think the metrics you are using are coming from node exporter. Did you specify any scrape targets in the constructor in Kafka?

**Comment by sed-i** [comment]
Confirmed, the labels are indeed missing from kafka and node exported (self metrics do have labels).
Interesting that over juju-info, node-exported did have labels.

```bash
$ curl -s localhost:12345/agent/api/v1/metrics/targets | jq
```
```json
{
  "status": "success",
  "data": [
    {
      "instance": "8a263e2cb5aa14d547a83233b446c3bb",
      "target_group": "integrations/agent",
      "endpoint": "http://127.0.0.1:12345/integrations/agent/metrics",
      "state": "up",
      "labels": {
        "agent_hostname": "juju-0728c2-4",
        "instance": "127.0.0.1:12345",
        "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_self-monitoring",
        "juju_application": "agent",
        "juju_charm": "grafana-agent",
        "juju_model": "default",
        "juju_model_uuid": "863a1762-9849-4724-8f72-e2288d0728c2",
        "juju_unit": "agent/9"
      },
      "discovered_labels": {
        "__address__": "127.0.0.1:12345",
        "__metrics_path__": "/integrations/agent/metrics",
        "__scheme__": "http",
        "__scrape_interval__": "1m",
        "__scrape_timeout__": "10s",
        "agent_hostname": "juju-0728c2-4",
        "job": "integrations/agent"
      },
      "last_scrape": "2023-03-07T22:56:57.4414656Z",
      "scrape_duration_ms": 2,
      "scrape_error": ""
    },
    {
      "instance": "8a263e2cb5aa14d547a83233b446c3bb",
      "target_group": "integrations/node_exporter",
      "endpoint": "http://127.0.0.1:12345/integrations/node_exporter/metrics",
      "state": "up",
      "labels": {
        "agent_hostname": "juju-0728c2-4",
        "instance": "127.0.0.1:12345",
        "job": "juju_default_863a1762-9849-4724-8f72-e2288d0728c2_agent_node-exporter"
      },
      "discovered_labels": {
        "__address__": "127.0.0.1:12345",
        "__metrics_path__": "/integrations/node_exporter/metrics",
        "__scheme__": "http",
        "__scrape_interval__": "1m",
        "__scrape_timeout__": "10s",
        "agent_hostname": "juju-0728c2-4",
        "job": "integrations/node_exporter"
      },
      "last_scrape": "2023-03-07T22:57:07.842916768Z",
      "scrape_duration_ms": 6,
      "scrape_error": ""
    },
    {
      "instance": "8a263e2cb5aa14d547a83233b446c3bb",
      "target_group": "kafka_0",
      "endpoint": "http://localhost:9100/metrics",
      "state": "up",
      "labels": {
        "instance": "localhost:9100",
        "job": "kafka_0"
      },
      "discovered_labels": {
        "__address__": "localhost:9100",
        "__metrics_path__": "/metrics",
        "__scheme__": "http",
        "__scrape_interval__": "1m",
        "__scrape_timeout__": "10s",
        "job": "kafka_0"
      },
      "last_scrape": "2023-03-07T22:57:44.515858825Z",
      "scrape_duration_ms": 8,
      "scrape_error": ""
    },
    {
      "instance": "8a263e2cb5aa14d547a83233b446c3bb",
      "target_group": "kafka_1",
      "endpoint": "http://localhost:9101/metrics",
      "state": "down",
      "labels": {
        "instance": "localhost:9101",
        "job": "kafka_1"
      },
      "discovered_labels": {
        "__address__": "localhost:9101",
        "__metrics_path__": "/metrics",
        "__scheme__": "http",
        "__scrape_interval__": "1m",
        "__scrape_timeout__": "10s",
        "job": "kafka_1"
      },
      "last_scrape": "2023-03-07T22:56:54.162546704Z",
      "scrape_duration_ms": 0,
      "scrape_error": "Get \"http://localhost:9101/metrics\": dial tcp 127.0.0.1:9101: connect: connection refused"
    }
  ]
}
```

**Comment by dstathis** [comment]
https://github.com/canonical/grafana-agent-k8s-operator/blob/3c98895cf43d131d7fa6928e01221990f6ac85c7/src/machine_charm.py#L110-L112

I think this needs to be something like

```
    def metrics_jobs(self) -> list:
        """Return a list of metrics scrape jobs."""
        jobs self._cos.metrics_jobs
        for job in jobs:
            job["labels"] = self._principal_labels
        return jobs
```

**Comment by sed-i** [comment]
Fixed by #120.
